src/git-repository-async.js

```diff
@@ -0,0 +1,870 @@
+  }
+
+  // Refresh the cached status.
+  //
+  // Returns :: Promise<null>
+  _refreshStatus () {
+    this._refreshingCount++
+
+    return this.repoPromise
+      .then(repo => repo.getStatus())
+      .then(statuses => {
+        // update the status cache
+        const statusPairs = statuses.map(status => [status.path(), status.statusBit()])
+        return Promise.all(statusPairs)
+          .then(statusesByPath => _.object(statusesByPath))
+      })
+      .then(newPathStatusCache => {
+        if (!_.isEqual(this.pathStatusCache, newPathStatusCache) && this.emitter != null) {
+          this.emitter.emit('did-change-statuses')
+        }
+        this.pathStatusCache = newPathStatusCache
+        return newPathStatusCache
+      })
+      .then(_ => this._refreshingCount--)
+  }
+
+  // Refreshes the git status.
+  //
+  // Returns :: Promise<null>
+  //            Resolves when refresh has completed.
+  refreshStatus () {
+    // TODO add submodule tracking
+
+    const status = this._refreshStatus()
+    const branch = this._refreshBranch()
+    const aheadBehind = branch.then(branchName => this._refreshAheadBehindCount(branchName))
+
+    return Promise.all([status, branch, aheadBehind]).then(_ => null)
+  }
+
+  // Get the NodeGit repository for the given path.
+  //
+  // path    :: Optional<String>
+  //            The path within the repository. This is only needed if you want
+  //            to get the repository for that path if it is a submodule.
+  //
+  // Returns :: Promise<NodeGit.Repository>
+  getRepo (_path) {
+    if (this._destroyed()) {
+      return Promise.reject(new Error('Repository has been destroyed'))
+    }
+
+    if (!_path) return this.repoPromise
+
+    return this.isSubmodule(_path)
+      .then(isSubmodule => {
+        if (isSubmodule) {
+          return Git.Repository.open(_path)
+        } else {
+          return this.repoPromise
+        }
+      })
+  }
+
+  // Section: Private
+  // ================
+
+  // Is the repository currently refreshing its status?
+  //
+  // Returns :: Bool
+  _isRefreshing () {
+    return this._refreshingCount === 0
+  }
+
+  // Has the repository been destroyed?
+  //
+  // Returns :: Bool
+  _destroyed () {
+    return this.repoPromise == null
+  }
+
+  // Subscribe to events on the given buffer.
+  subscribeToBuffer (buffer) {
+    const bufferSubscriptions = new CompositeDisposable()
+
+    const refreshStatusForBuffer = () => {
+      const _path = buffer.getPath()
+      if (_path) {
+        this.refreshStatusForPath(_path)
+      }
+    }
+
+    bufferSubscriptions.add(
+      buffer.onDidSave(refreshStatusForBuffer),
+      buffer.onDidReload(refreshStatusForBuffer),
+      buffer.onDidChangePath(refreshStatusForBuffer),
+      buffer.onDidDestroy(() => {
+        bufferSubscriptions.dispose()
+        this.subscriptions.remove(bufferSubscriptions)
+      })
+    )
+
+    this.subscriptions.add(bufferSubscriptions)
+    return
```

Comment:
I don't think this explicit `return` is needed since it is the last line of this function.